regression_tests/utils.go

```diff
@@ -0,0 +1,222 @@
+	"time"
+
+	walletclient "github.com/bitcoin-sv/spv-wallet-go-client"
+	"github.com/bitcoin-sv/spv-wallet-go-client/xpriv"
+	"github.com/bitcoin-sv/spv-wallet/models"
+	"github.com/bitcoin-sv/spv-wallet/models/filter"
+)
+
+const (
+	atSign                   = "@"
+	domainPrefix             = "https://"
+	domainSuffixSharedConfig = "/v1/shared-config"
+
+	ClientOneURLEnvVar         = "CLIENT_ONE_URL"
+	ClientTwoURLEnvVar         = "CLIENT_TWO_URL"
+	ClientOneLeaderXPrivEnvVar = "CLIENT_ONE_LEADER_XPRIV"
+	ClientTwoLeaderXPrivEnvVar = "CLIENT_TWO_LEADER_XPRIV"
+
+	timeoutDuration = 30 * time.Second
+)
+
+var (
+	explicitHTTPURLRegex = regexp.MustCompile(`^https?://`)
+	envVariableError     = errors.New("missing xpriv variables")
+)
+
+type regressionTestUser struct {
+	XPriv   string `json:"xpriv"`
+	XPub    string `json:"xpub"`
+	Paymail string `json:"paymail"`
+}
+
+type regressionTestConfig struct {
+	ClientOneURL         string
+	ClientTwoURL         string
+	ClientOneLeaderXPriv string
+	ClientTwoLeaderXPriv string
+}
+
+// getEnvVariables retrieves the environment variables needed for the regression tests.
+func getEnvVariables() (*regressionTestConfig, error) {
+	rtConfig := regressionTestConfig{
+		ClientOneURL:         os.Getenv(ClientOneURLEnvVar),
+		ClientTwoURL:         os.Getenv(ClientTwoURLEnvVar),
+		ClientOneLeaderXPriv: os.Getenv(ClientOneLeaderXPrivEnvVar),
+		ClientTwoLeaderXPriv: os.Getenv(ClientTwoLeaderXPrivEnvVar),
+	}
+
+	if rtConfig.ClientOneLeaderXPriv == "" || rtConfig.ClientTwoLeaderXPriv == "" {
+		return nil, envVariableError
+	}
+	if rtConfig.ClientOneURL == "" || rtConfig.ClientTwoURL == "" {
+		rtConfig.ClientOneURL = "http://localhost:3003"
+		rtConfig.ClientTwoURL = "http://localhost:3003"
+	}
+	return &rtConfig, nil
+}
+
+// getSharedConfig retrieves the shared configuration from the SPV Wallet.
+func getSharedConfig(xpub string, clientUrl string) (*models.SharedConfig, error) {
+	req, err := http.NewRequest(http.MethodGet, clientUrl+domainSuffixSharedConfig, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set(models.AuthHeader, xpub)
+	client := http.Client{
+		Timeout: timeoutDuration,
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("failed to get shared config: %s", resp.Status)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	var configResponse models.SharedConfig
+	if err := json.Unmarshal(body, &configResponse); err != nil {
+		return nil, err
+	}
+
+	if len(configResponse.PaymailDomains) != 1 {
+		return nil, fmt.Errorf("expected 1 paymail domain, got %d", len(configResponse.PaymailDomains))
+	}
+	return &configResponse, nil
+}
+
+// createUser creates a set of keys and new paymail in the SPV Wallet.
+func createUser(paymail string, paymailDomain string, instanceUrl string, adminXPriv string) (*regressionTestUser, error) {
```

Comment:
can not u pass here ctx in upper function, as each of those helper functions are containing `context.Background()`, so i think would be better to create ctx once in TestRegression and then pass it to these functions

Reply:
Moved to upper function.